frb_codegen/src/commands.rs

```diff
@@ -245,3 +250,111 @@ pub fn build_runner(dart_root: &str) -> Result {
     }
     Ok(())
 }
+
+lazy_static! {
+    static ref CACHE: Mutex<HashMap<String, String>> = Mutex::new(HashMap::new());
+}
+
+pub fn cargo_expand(dir: &str, module: Option<String>, file: &str) -> String {
+    let manifest_dir = env::var("CARGO_MANIFEST_DIR").unwrap_or_default();
+    #[cfg(windows)]
+    let manifest_dir = &manifest_dir.replace('\\', "/");
+    if !manifest_dir.is_empty() && dir == manifest_dir {
+        warn!(
+            "can not run cargo expand on {dir} because cargo is already running and would block cargo-expand. This might cause errors if your api contains macros."
+        );
+        return fs::read_to_string(file).unwrap_or_default();
+    }
+    let mut cache = CACHE.lock().unwrap();
+    let expanded = cache
+        .entry(String::from(dir))
+        .or_insert_with(|| run_cargo_expand(dir));
+    extract_module(expanded, module)
+}
+
+fn extract_module(mut expanded: &str, module: Option<String>) -> String {
+    if let Some(module) = module {
+        let mut spaces = 0;
+        for module in module.split("::") {
+            let searched = format!("mod {module} {{\n");
+            let start = expanded
+                .find(&searched)
+                .map(|n| n + searched.len())
+                .unwrap_or_default();
+            if start == 0 {
+                continue;
+            }
+            let end = expanded[start..]
+                .find(&format!("\n{}}}", " ".repeat(spaces)))
+                .map(|n| n + start)
+                .unwrap_or(expanded.len());
+            expanded = &expanded[start..end];
```

Comment:
hmm shall we use such a mutable variable (`mut expanded: &str`)? it may be great to make it immutable when possible

Reply:
i changed it to be immutable using fold. I am not sure if that is better. I think it's fine but it's a bit more complicated to read.

Reply:
Hmm, I am not sure. Originally I thought about making the param immutable and making a variable like `let mut interestd_part = expanded; ... change interested_part ...` on your old code. But anyway both approaches looks OK.